044-kraft-liveness-readiness.md

```diff
@@ -0,0 +1,174 @@
+
+The other possible value is 127 which represents UNKNOWN.
+
+You can see how the state transitions in the [BrokerLifecycleManager](https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/server/BrokerLifecycleManager.scala#L379), specifically that 
+the broker transitions from STARTING to RECOVERY once it has caught up with the cluster metadata, then it transitions to RUNNING as long as it is not fenced.
+
+The broker nodes cannot catch up with the cluster metadata until the controller quorum has been formed, so they will not move out of the STARTING state until the quorum leader election has happened.
+
+## Proposal
+
+There are three different types of liveness/readiness probe that we need to consider, first for a controller only, then 
+for a broker only, and finally for a combined node that has both roles.
+
+The following statements describe the intent of the proposed probes:
+
+* A controller is:
+    * "alive" if it has a process running.
+    * "ready" when it is ready to join the controller quorum.
+* A broker is:
+    * "alive" if it has a process running.
+    * "ready" when it is ready to start accepting producer/consumer requests.
+* A combined node is:
+    * "alive" if it has a process running.
+    * "ready" when it is ready to start accepting producer/consumer requests.
+        * This implies that the controller quorum will have been formed while the nodes are not marked as ready, since it cannot start accepting producer/consumer requests until it can talk to the leader.
+        * This relies on the node still accepting incoming connections from other controller nodes even if it isn't actually marked as "ready" yet.
+
+### Controller only mode
+
+The proposed probes are:
+
+* Liveness: controller is listening on the port of the first address in `controller.listener.names` (9090)
+* Readiness: controller is listening on the port of the first address in `controller.listener.names` (9090)
+
+### Broker only mode
+
+The proposed probes are:
+
+* Liveness: broker is listening on the port of the address of `inter.broker.listener.name` (9091)
+* Readiness: the BrokerState metric >=3 && != 127
+
+**Note:** This means the brokers will not become ready until a majority of the controllers 
+are up and running.
+This is similar to the current behaviour of brokers when ZooKeeper is not ready.
+
+### Combined mode
+
+The proposed probes are:
+
+* Liveness: node is listening on the port of the first address in `controller.listener.names` (9090)
+* Readiness: the BrokerState metric >=3 && != 127
+
+**Note:** This means the nodes will not become ready until a majority of the other controllers are up and running.
+This is acceptable because the Strimzi headless services use "publishNotReadyAddresses", which 
+means the nodes will be able to communicate even if they are currently marked as not ready.
+
+### Impact on the KafkaRoller
```

Comment:
From here on, the proposal seems to be quite confusing in terms of what is proposes, what is being approved and what should be done in the future. I think that need to be clarified and we need to make sure there is a clear deliverable (or make it clear that there is no real deliverable).

It seems to me that, this proposal:
* Suggests how the liveness and readiness should be implemented in phase 4
* Leaves the phase 3 completely unspecified, so essentially this proposal cannot be done until the phase 4 is specified in some proposal and approved?
* The content of the `Impact on the KafkaRoller` chapter makes sense in general, but it seems IMHO too simplified to take it as something to be approved. So I think it should have some clear note that this will be covered in detail in another proposal? (at least that is what I assume?)

----

To be clear, I do not have problem with having this agreed upon before the other parts are done. It should be just more clear what are we agreeing upon and make it clear what is just outline of something what will be discussed and approved in separate proposal (e.g. the `Impact on the KafkaRoller` part).

Reply:
I see what you mean, I've updated the wording to try to make it clearer what is part of this proposal and what parts are more for information to help inform a separate proposal for KafkaRoller in KRaft mode. I've also merged phase 3 and 4 below, since the KafkaRoller changes wouldn't directly be covered by this proposal, it is just required before the final changes can be made to the liveness and readiness checks. Hopefully that helps clarify.